### DISCUSSION TITLE
MRIReco.jl Community Meetings

### DISCUSSION THREAD

**Discussion by mrikasper** [MagneticResonanceImaging/MRIReco.jl#99] [General]
Hi everyone,

Thank you all for the great meeting today. We decided to post updates/minutes/next meeting infos in this new discussion.

In general, we will meet

Every **first Thursday** of the month
at **12:00 UTC** (08:00 EST / 14:00 CET)

So the next meeting will be 

_Thu, Oct, 6th at 12:00 UTC_

The idea is to develop a concrete agenda/discussion points before each meeting in this thread.

**Comment by mrikasper** [comment]
I am happy to share some meeting notes I took today, please add and amend:

## Minutes Meeting 01 (2022-09-01)

### Discussion Points
1.	`MRIReco` Philosophy (Tobias): common task in common packages (MRI, MPI…), encapsulation
a.	e.g., regularized least squares, NFFT in extra packages
b.	Tobias: develop algorithmic stuff in max abstract way (reuse in other non-MRI modalities)
c.	Goal: people can realize their research! bring issues (e.g., _I have trouble reconstructing that data!_)
2.	`MRIReco` Vision (Tobias): Understand `MRIReco` as middle layer
a.	between MRI-specific layer and concrete operators
b.	role of Mirco and Tobias: not so much MRI in the last year
c.	if there are issues, please report them (but resources limited)
d.     user engagement necessary, for bigger feature: collaboration! (see below for deep learning)
3.	`MRILib` (Tobias): so far only an idea for the highest layer Everything in MRI (recon, simulator, files)
a.	like a Matlab Toolbox
b.	MRIReco should not be the leaf package
4.	Documentation (Lars)
a.	Best way to make high level 
b.	run Julia code with doctests (Tobias)
c.	or `Literate.jl`, e.g,. [Jeff's example](https://juliaimagerecon.github.io/ImagePhantoms.jl/stable/generated/examples/10-mri-sense/)
d.	_TODO (Lars)_: 3D stack of spirals with Alejandro!
5.	Real-time online reconstruction (Tobias/Aurelien)
a.	Tobias: with Julia on MPI scanner!
b.	Aurelien in Bordeaux wants to do it
6.	GPU support (Tobias/Mirco)
a.	trial for NFFT-based operators
b.	Do we need it?: AMD 256 threads very fast (NFFT 2-3x faster with GPU)
c.      at which level? NFFT? larger algorithms?
d.	MIRT user requests it (Jeff)
e.	BART supports it (Aurelien)
7.	Development of Deep Learning (Tobias and Alexander)
a.      currently trying machine-learning based image reconstruction via Julia tools (Flux)…pick up in MRIReco?
b.	First step: Adjoint operators for chain rule
c.	_TODO (Alexander/Tobias?)_: making MRIReco AD-friendly (for joint estimation)

### Issues
1.	General Issue reporting  
a.	Minimal working example (mandatory!)
b.	Did you have this scenario with another software (e.g., BART)?
c.	if confident in Julia, try to propose fix yourself (via PR), or within issue discussion do it with Mirco/Tobias
2.	Test not only breaking changes, but also Performance regression 
a.      Is code getting slower? e.g., via "`BenchmarkTools`"
b.	Jeff: to speed up CI, could be a workflow (Benchmark CI) flag, runs only when performance evaluation!
a.	Expand on `MRIRecoBenchmarks`: relative benchmarking to other packages (accuracy and benchmark time)
3.	Memory (Alejandro: 24 slices, 32 ch, 3D stack of spiral)
a.	maybe conversion to Float64 happens somewhere?
4.	BART vs MRIReco.jl: slow high-level functions (Aurelien)
a.	type instabilities!
b.	in the high-level functions, type-unstable
c.	e.g., read a file and don't know what it contains, but want to dispatch (dynamic dispatch!), then further down in stack, make it type-stable
5.	Use-case specific tests:
a.	Create issue for Tobias for test data (only he can upload data (artifacts) to CI)
b.	Make sure you are able  to share the data (phantom?): dummy data, small!
6.	Subfolders as extra packages
a.	to do integration test for the combined repo
b.	e.g., MRIBase is in, but not NFFT
c.	navigate many repos: Documentation, continuous integration setup for all of it
d.	sub-package not distributed on its own usually
e.	one advantage: subpackage will be checked out on disk into same location (same repo)
f.	but still have to individually `dev MRIReco, MRIBase, MRIFiles`
g.	in general (not subpackage-related): version numbers have to be released in order for things to work!
h.	two ways to work with `MRIReco`: as user add latest release; as developer `dev` it and all sub-packages

### Valuable Resources
1.	https://github.com/JeffFessler/MIRT.jl/blob/main/.github/workflows/BenchmarkCI.yml
2.	https://github.com/JeffFessler/MIRT.jl/tree/main/benchmark
3.	https://github.com/JakobAsslaender/BartIO.jl
4.	https://atrotier.github.io/EDUC_JULIA_CS_MP2RAGE/
5.	https://github.com/cncastillo/KomaMRI.jl/blob/master/Project.toml

**Comment by mrikasper** [comment]
> Hi everyone,
> 
> Thank you all for the great meeting today. We decided to post updates/minutes/next meeting infos in this new discussion.
> 
> In general, we will meet
> 
> Every **first Thursday** of the month at **12:00 UTC** (08:00 EST / 14:00 CET)
> 
> So the next meeting will be
> 
> _Thu, Oct, 6th at 12:00 UTC_
> 
> The idea is to develop a concrete agenda/discussion points before each meeting in this thread.

Dear Julia MRIReco enthusiasts (@tknopp @migrosser @JeffFessler @aTrotier @JakobAsslaender @alexjaffray et al.),

Are we still good for meeting tomorrow? Or do people prefer different dynamics given the quite lively written discussions that already happen around the issues and enhancements of MRIReco.jl?

Together with @alexjaffray, we were wondering about using the `MagneticResonanceImaging` organization for an upcoming publication related to [this abstract](https://archive.ismrm.org/2022/2435.html)

1. GitHub organization: @tknopp mentioned the vision of having a hub for Julia-MR code in `MagneticResonanceImaging` with joint documentation. We are interested to try for the code related to the ISMRM abstract. What are practical steps for that (we just started with our own `Documenter.jl` usage)?


All the best,
Lars

**Reply by tknopp** [reply]
Sure, the meeting is scheduled. We need to use another link, since I am not a host of the other one and cannot start it:
  https://tuhh.zoom.us/j/83981707416?pwd=dzJOOStoV1MySE14c0FWK3dSbmtTUT09
Tomorrow I have just one hour and have to leave then.

**Comment by mrikasper** [comment]
Here are @aTrotier topics from the the other discussion #82 


> - Benchmark results
> - Accuracy of fista (example in benchmark or https://atrotier.github.io/EDUC_JULIA_CS_MP2RAGE/)
> - Reconstruction of repetitions : PR #105 
> - Conversion from acq to BartIO and BartIO to acq ? (also part of PR #105)
> - Be able to use different trajectories for all dimensions (only implemented for echoes)
> - ISMRM abstracts ?

**Reply by aTrotier** [reply]
- Documentation enhancement lists : https://github.com/MagneticResonanceImaging/MRIReco.jl/discussions/104

**Comment by JakobAsslaender** [comment]
I am joining tomorrow--sorry for missing last month's meeting! 
I have two items that I would like to discuss if there is time:
- decision on whether to move ESPIRIT code (and other related methods) to a separate `MRICoilSensitivities.jl` packages (https://github.com/MagneticResonanceImaging/MRIReco.jl/issues/109)
- discussion on whether to pull the back projection out of the algorithms in [RegularizedLeastSquares.jl](https://github.com/tknopp/RegularizedLeastSquares.jl). This would have three advantages:
   1. same interface as `CG`
   2. one can precompute the back projection and delete the data before starting the actual reconstruction to save memory (this makes an actual difference in our use-case)
   3. it is sufficient implement the back projection operator (`AH`) and the normal (`AH*A`) operator (e.g., by using the Toeplitz trick) w/o having to fake the forward operator `A` that is not required for the reconstruction.

**Comment by aTrotier** [comment]
Hi, 
I won't be able to meet you tomorrow but I can share with you what are my plans for next month :

- Coil compression :
    - Add geometric coil compression (for ecc I don't know how it works I guess it is highly related to espirit maybe @JakobAsslaender has some thoughts about
    - Create a notebook to compare vcc vs gcc 
    - For latter work on a non-cartesian implementation
- Implement others way to perform discrete wavelet decomposition and find a way to benchmark them (see topics : https://github.com/tknopp/SparsityOperators.jl/issues/12 and https://github.com/JuliaDSP/Wavelets.jl/issues/80)
   - zero-padding to the next 2^n and use 2D/3D dwt
   - Fully separable : perform dwt along dim 1 to the maximum level then along dim 2
   - mix -> 2D/3D dwt to the maximum level then for the perform 1D DWT along each remaining dimension

**Comment by tknopp** [comment]
Hi, I will again be in the Zoom meeting today (Thu, Nov, 3th at 12:00 UTC):

https://tuhh.zoom.us/j/83981707416?pwd=dzJOOStoV1MySE14c0FWK3dSbmtTUT09

I can give an overview about the new package structure and outline things that happened in the last month.

**Comment by JakobAsslaender** [comment]
Quick question to @tknopp and everyone: Is there any chance we could push this meeting a couple of hours? For east coast US, the 12pm UTC is 7am, which is way before my time, and I think @alexjaffray is on the west coast, where this 4am. Do we have anyone in the asian/pacific area that is joining this meeting for whom a later time would not work? My first suggestion would be 4pm UTC, which would put everyone in Europe and North America between 8am and 5pm, i.e., in "normal" working hours.

**Reply by tknopp** [reply]
Yes, of course we can shift to 4pm. Unfortunately, I don't have time next Thursday. If you want to still do the meeting someone needs to provide a Zoom Link. I can join in January again.

**Reply by alexjaffray** [reply]
Thanks @JakobAsslaender for proposing this, It would be great to shift, as 5am my time is doable but 4am is starting to be quite early...

**Reply by alexjaffray** [reply]
Would anyone be able to provide a Zoom link for the meeting scheduled for tomorrow? I would offer but my meeting links on Zoom are limited to 40 mins.  @aTrotier you mentioned you had some things to discuss this month since you couldn't make it last time. 

An aside, looking at my calendar I also realize now that the new proposed time conflicts with the scheduled BART webinar on recon this week (also 8am PST). Should we postpone our next meeting until January so that @tknopp joins us (and we avoid this conflict?)

**Reply by JeffFessler** [reply]
I suggest waiting to Jan so we can plan the zoom a bit further ahead...

**Reply by aTrotier** [reply]
Agreed :)

**Reply by tknopp** [reply]
Hey all, in theory we would today have the next MRIReco.jl community call. However, I am right now quite busy and would therefore suggest that we instead use Github for interactions.

**Comment by aTrotier** [comment]
Sure, anyway I was not able to work a lot on MRIReco during the last 2
months.
I need to work on the fista accuracy issue / wavelet implementation

Le jeu. 5 janv. 2023 à 10:21, Tobias Knopp ***@***.***> a
écrit :

> Hey all, in theory we would today have the next MRIReco.jl community call.
> However, I am right now quite busy and would therefore suggest that we
> instead use Github for interactions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MagneticResonanceImaging/MRIReco.jl/discussions/99#discussioncomment-4599623>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC5P7O7KBVSEYMW2YNIHXGTWQ2HCPANCNFSM6AAAAAAQCMZ54E>
> .
> You are receiving this because you were mentioned.Message ID:
> <MagneticResonanceImaging/MRIReco.jl/repo-discussions/99/comments/4599623@
> github.com>
>


-- 
------------------------------------------------------------
TROTIER Aurélien
31 allée de la Biotte
33470 Gujan-Mestras
06.09.12.61.61
***@***.***

**Comment by JeffFessler** [comment]
Does anyone have a Julia-related presentation at ISMRM 2023?  If so, maybe it's a convenient place to meet up and say hello?

**Comment by JakobAsslaender** [comment]
We have no related presentation, but I would love the opportunity to say hello!

**Comment by mrikasper** [comment]
We have two posters (similar to the ones in Sedona) on spiral diffusion imaging with MRIReco.jl et al. :-)

1. Wednesday, 07 June 2023 | 08:15 - 09:15
Session Title
[DTI & DWI I](https://submissions.mirasmart.com/ISMRM2023/Itinerary/EventDetail.aspx?evt=225)
Acquisition & Analysis
Feasibility of diffusion imaging using SMS-spiral acquisition with corrections on gradient waveform and field inhomogeneities
Zhe Wu
Program Number: 3614
Computer Number: 30
Presenter: Zhe Wu ("Tim"), BRAIN-TO Lab, Techna Institute, University Health Network 

2. Wednesday, 07 June 2023 | 14:30 - 15:30
Session Title
[Diffusion: Applications](https://submissions.mirasmart.com/ISMRM2023/Itinerary/EventDetail.aspx?evt=349)
Contrast Mechanisms
Feasibility of spiral diffusion imaging on a clinical 3T MR system
Program Number: 4164
Computer Number: 76
Presenter: Lars Kasper, BRAIN-TO Lab, Techna Institute, University Health Network

**Reply by aTrotier** [reply]
Are you all ok for a meeting during :
```
Wednesday, 07 June 2023 | 14:30 - 15:30
Session Title
[Diffusion: Applications](https://submissions.mirasmart.com/ISMRM2023/Itinerary/EventDetail.aspx?evt=349)
Contrast Mechanisms
Feasibility of spiral diffusion imaging on a clinical 3T MR system
Program Number: 4164
Computer Number: 76
Presenter: Lars Kasper, BRAIN-TO Lab, Techna Institute, University Health Network
```

**Reply by alexjaffray** [reply]
I have another abstract during this time but am at a very close computer (93) so will drop by!

**Comment by aTrotier** [comment]
I also have one digital poster but I am mostly using MRIReco.jl to convert Bruker rawdata with BartIO.jl for the reconstruction :

Thursday, 08 June 2023 | 08:15 - 09:15
Session Title
[Image Reconstruction Methods I](https://submissions.mirasmart.com/ISMRM2023/Itinerary/EventDetail.aspx?evt=310)
Acquisition & Analysis
Accelerated 3D Multi-Echo Spin-Echo sequence for T2 mapping of mouse brain metastases using a subspace-based reconstruction
Program Number: 4624
Computer Number: 29

**Comment by tknopp** [comment]
I am not at ISMRM this year unfortunately.
